File: Python/BestTimeToBuyAndSellStock.py
def maxProfit(prices):
    """
    :type prices: List[int]
    :rtype: int
    """
    def min(arr):
        index = 0
        min = arr[index]
        for i in range(0, len(arr)):
            if(arr[i] < min):
                index = i
                min = arr[i]
            else:
                pass
        return [index, min]

    def max(arr, index):
        max = arr[index]
        for i in range(index, len(arr)):
            if (arr[i] >= max):
                index = i
                max = arr[i]
            else:
                pass
        return [index, max]

    print(min([2, 4, 1]))

    def trade(arr):
        # return max(arr, min(arr)[0]), min(arr)
        if (max(arr, min(arr)[0])[0] > min(arr)[0]):
            return max(arr, min(arr)[0])[1] - min(arr)[1]
        else:
            return 0
    return trade(prices)


print(maxProfit([2, 4, 1]))

File: JavaScript/BestTimeToBuyAndSellStock.js
// You are given an array prices where prices[i] is the price of a given stock on the ith day.

// You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.

// Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.

/**
 * @param {number[]} prices
 * @return {number}
 */


const maxProfit = (prices) => {

    let maxValue = Math.max(...prices);
    let profit = 0;

    for (let i = 0; i < prices.length; i++) {
        if (prices[i] < maxValue) {
            maxValue = prices[i];
        }
        else if (prices[i] - maxValue > profit) {
            profit = prices[i] - maxValue;
        }
    }
    return profit;
};


console.log(maxProfit([7, 1, 5, 3, 6, 4]));